### ISSUE TITLE
Rendering Multiple Elements in a slot

### ISSUE THREAD

**Issue by yagudaev** [github/view_component#403]
### Steps to reproduce
Consider a modal with a foot that looks like this:

<img width="570" alt="Screen Shot 2020-07-14 at 11 36 41 AM" src="https://user-images.githubusercontent.com/1386966/87463504-5ba5a400-c5c6-11ea-88eb-ebeb2717e01a.png">

The component preview looks like this:

```ruby
class ModalComponentPreview < ViewComponent::Preview
  def default
    render(ModalComponent.new) do |component|
      component.slot(:modal_body) { Faker::Quote.matz }
      component.slot(:modal_footer) do
        component.render(ButtonComponent.new(type: 'default', data: { action: 'click->modal-component#close'})) { 'Cancel' },
        component.render(ButtonComponent.new(type: 'main', classes: 'ml-3')) { 'Save' }
      end
    end
  end
end
```

### Expected behavior
Both buttons show up as in the screenshot above.

### Actual behavior
Only one button shows up, the save button.

The reason for this is that render evaluates to a string and only the last expression is returned from the block. Therefore only the last render will show up.

One way I find to get it to work is:

```ruby
      component.slot(:modal_footer) do
        [
          component.render(ButtonComponent.new(type: 'default', data: { action: 'click->modal-component#close'})) { 'Cancel' },
          component.render(ButtonComponent.new(type: 'main', classes: 'ml-3')) { 'Save' }
        ].join.html_safe
      end
```

Is there a better way? Can we document the recommended way?

Finally, can we do better than this in terms of an API?

### System configuration
All erbs

**Comment by joelhawksley** [comment]
@yagudaev thanks for taking the time to file an issue! Your best bet is to use preview templates, recently added by @juanmanuelramallo: https://github.com/github/view_component#preview-templates

**Comment by scottbarrow** [comment]
@joelhawksley this isn't specific to Previews, the issue I believe being reported is that you cannot pass a block of components or elements to a slot block and therefore the block is required to concatenate the contents of the block with join.

The alternative is to pass multiple blocks as per the documentation
```ruby
render(ModalComponent.new) do |component|
      component.slot(:modal_body) { Faker::Quote.matz }
      component.slot(:modal_footer) do
        component.render(ButtonComponent.new(type: 'default', data: { action: 'click->modal-component#close'})) { 'Cancel' },
      end        
      component.slot(:modal_footer) do
        component.render(ButtonComponent.new(type: 'main', classes: 'ml-3')) { 'Save' }
      end    
    end
 ```

but thats super verbose. It would be great if you could pass multiple sub-components to a slot block

**Comment by joelhawksley** [comment]
@scottbarrow Generally, we would do the following in ERB:

```erb
<%= render(ModalComponent.new) do |component|
  <% component.slot(:modal_body) do %>
    <%= Faker::Quote.matz %>
  <% end %>
  <% component.slot(:modal_footer) do %>
    <%= render(ButtonComponent.new(type: 'default', data: { action: 'click->modal-component#close'})) { 'Cancel' } %>
    <%= render(ButtonComponent.new(type: 'main', classes: 'ml-3')) { 'Save' } %>
  <% end %>
<% end %>
```

I believe the issue you're having is just as much an issue with rendering a partial from inside a method definition, such as in a controller. I'm all for making previews more flexible, but am not sure what the right path forward is here. Any ideas @juanmanuelramallo ?

**Comment by yagudaev** [comment]
@joelhawksley yep that ERB path does work well.

I was just trying to render the whole component in a single ERB block, which I guess is not the recommended way.

Preview Templates do solve the problem in a different way. I've used them for this:

![Screen Shot 2020-07-16 at 2 06 31 PM](https://user-images.githubusercontent.com/1386966/87722807-9bf35680-c76d-11ea-9bec-820fd84422b7.png)

which I just created a 

```html
# ../component/previews/status_badge_component_preview/default.html.erb

<div class="mb-2">
  <%= render(StatusBadgeComponent.new(status: :active)) %>
  <%= render(StatusBadgeComponent.new(status: :accepted)) %>
</div>
<div class="mb-2">
  <%= render(StatusBadgeComponent.new(status: :default)) %>
  <%= render(StatusBadgeComponent.new(status: :inactive)) %>
</div>
<div class="mb-2">
  <%= render(StatusBadgeComponent.new(status: :suspended)) %>
</div>

```

By my original instinct was to have something like this:

```ruby
class StatusBadgeComponentPreview < ViewComponent::Preview
  def default
    tag.div class: 'mb-2' do
      render(StatusBadgeComponent.new(status: :active))
      render(StatusBadgeComponent.new(status: :accepted))
    end

    tag.div class: 'mb-2' do
      render(StatusBadgeComponent.new(status: :default))
      render(StatusBadgeComponent.new(status: :inactive))
    end

    tag.div class: 'mb-2' do
      render(StatusBadgeComponent.new(status: :suspended))
    end
  end
end
```

I think the way to do that is to have all the ActiveView methods be added to and output buffer.

I believe ActiveAdmin does some of that, but it fall flat in a lot of cases and would have similar issues to this one.

In summary, the workarounds you provide here work great. But the more ergonomic interface would be just inlining the HTML into the controller. Feel free to close this issue, I think tackling this will require a different ticket and research around how ActiveAdmin and other ruby libraries allow writing HTML right in the controller.

Maybe we will even get RBX - like JSX but a million times better, because it will be ruby... if PHP can do it (facebook's pre-react coding library), so can Ruby

**Comment by joelhawksley** [comment]
@yagudaev thanks for following up! I'm cool to leave this issue open. I'd love to see us pursue a solution more along the lines of what you're proposing. Would you like to give it a shot?

**Comment by yagudaev** [comment]
@joelhawksley yeah I can take a look at how we might want to do that, but I'll need some help.

Looking at ActiveAdmin you write code like:

```ruby
tabs do
  tab :all do
    render 'posts', collection: Posts.all
  end

  tab :active do
    render 'posts', collection: Posts.active
  end

  tab :inactive do
    render 'posts', collection: Posts.inactive
  end
end
```

Tabs and Tab each map to a factory method that creates the TabsComponent and TabComponent classes. It will be interesting to try and do something similar that you can mix into the component to allow something like that. Alternatively, it can go through a single artiry like `component.tabs` or `element.tabs` or even `ViewComponent.create('Tab', **args)`. React does something like that too: https://reactjs.org/docs/react-without-jsx.html.

We have so many cool API we can create with ruby through contexts and DSLs. Even

```
class ModalComponentPreview < ViewComponent::Preview
  def default
    view do
       modal do
          slot(:modal_body) { Faker::Quote.matz }
          slot(:modal_footer) do
            button(type: 'default', data: { action: 'click->modal-component#close'})) { 'Cancel' },
            button(type: 'main', classes: 'ml-3')) { 'Save' }
          end
      end
    end
  end
end
```

For the time being though, I think I'd like to keep it simple and enhance previews to support rendering multiple elements without resulting to template files.

**Comment by joelhawksley** [comment]
@yagudaev interesting. This kind of feature would also help us with rendering components from controllers, where it could be nice to avoid making a separate template in simple cases.
